Amount decimals validation

rcisterna opened the pull request:
Validations were implemented in the following methods:

- `WebPayNormal->initTransaction`
- `WebPayMallNormal->initTransaction`
- `WebpayOneClick->authorize`
- `WebpayCompleteTransaction->initCompleteTransaction`

Consider that before these changes, these methods didn't throw up exceptions, so please validate the implementation. Special consideration to `WebPayMallNormal`, where if one amount is invalid, no one is processed.

saengate commented:
@rcisterna This solves the problem in case the currency does not handle decimals, but what happens if the use of USD is enabled again?

gonzunigad commented:
Btw, I think it should be called ‘InvalidAmountException’ and include a
message to help the user understand what’s the problem.

On Mon, 6 Apr 2020 at 14:52 SaenGate <notifications@github.com> wrote:

> @rcisterna <https://github.com/rcisterna> This solves the problem in case
> the currency does not handle decimals, but what happens if the use of USD
> is enabled again?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/TransbankDevelopers/transbank-sdk-php/pull/122#issuecomment-609973797>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAINNBTGAKLCM65FGXWEFSLRLIQGVANCNFSM4MCPFJ2Q>
> .
>
-- 


GONZALO DE​-​SPIRITO

*FRESHWORK* STUDIO CHILE

​+569 8815 3776 | ​ <gonzalo@freshwork.cl>gonzalo@freshwork.cl

gonzunigad commented on lib/webpay/WebpayCompleteTransaction.php:
```diff
@@ -204,8 +207,17 @@ function _initCompleteTransaction($initCompleteTransaction) {
+    function initCompleteTransaction($amount, $buyOrder, $sessionId, $cardExpirationDate, $cvv, $cardNumber)
+    {
+        // validaciones $amount
+        if (!is_numeric($amount))
```
Please use PSR-2. 
In that case if statements shoukld always use brackets, even single line ones. :)

gonzunigad commented on lib/webpay/WebpayOneClick.php:
```diff
@@ -316,8 +319,17 @@ function finishInscription($token) {
+        if (!is_numeric($amount))
+            throw new AmountException();
+        if (is_float($amount))
+            throw new AmountDecimalsException();
```
```suggestion
            throw new AmountWithDecimalsException('Given amount has decimals. Webpay only accepts integer amounts. Please remove decimal places. ');
```

rcisterna replied:
shoud I use a new exception type, or just a different message?

gonzunigad commented on lib/webpay/WebpayOneClick.php:
```diff
@@ -316,8 +319,17 @@ function finishInscription($token) {
+    {
+        // validaciones $amount
+        if (!is_numeric($amount))
+            throw new AmountException();
```
```suggestion
            throw new InvalidAmountException('Given amount is not numeric');
```

gonzunigad commented on lib/webpay/WebpayOneClick.php:
```diff
@@ -316,8 +319,17 @@ function finishInscription($token) {
+        // validaciones $amount
+        if (!is_numeric($amount))
+            throw new AmountException();
+        if (is_float($amount))
```
is_float should not work as expected. Sometimes the amout could come from a user input or from a database query on a string variable. 
![image](https://user-images.githubusercontent.com/1103494/78599023-5b0aae00-781e-11ea-8d53-4ce34a5d7f0c.png)

```suggestion
        if (is_float($amount))
```

gonzunigad replied:
You yousl probably need something like this: 
![image](https://user-images.githubusercontent.com/1103494/78599186-ae7cfc00-781e-11ea-9131-7d4ab6c361d6.png)

rcisterna commented:
Just updated de branch with the @gonzunigad requests:
- Fixed amounts validation.
- Single exception (`InvalidAmountException`) with different messages depending on the validation.

@saengate the currency is not validated, because at least for now all transactions are in CLP, according to @afiebig.
